pkg/apis/core/validation/validation_test.go

```diff
@@ -12522,84 +12996,423 @@ func TestValidateServiceUpdate(t *testing.T) {
+			tweakSvc: func(oldSvc, newSvc *core.Service) {
+				oldSvc.Spec.Type = core.ServiceTypeExternalName
+				newSvc.Spec.Type = core.ServiceTypeClusterIP
+			},
+			numErrs: 0,
+		},
+		{
+			name: "invalid: convert to ExternalName",
+			tweakSvc: func(oldSvc, newSvc *core.Service) {
+				singleStack := core.IPFamilyPolicySingleStack
+
+				oldSvc.Spec.Type = core.ServiceTypeClusterIP
+				oldSvc.Spec.ClusterIP = "10.0.0.10"
+				oldSvc.Spec.ClusterIPs = []string{"10.0.0.10"}
+				oldSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
+				oldSvc.Spec.IPFamilyPolicy = &singleStack
+
+				newSvc.Spec.Type = core.ServiceTypeExternalName
+				newSvc.Spec.ExternalName = "foo"
+				/*
+					not removing these fields is a validation error
+					strategy takes care of resetting Families & Policy if ClusterIPs
+					were reset. But it does not get called in validation testing.
+				*/
+				newSvc.Spec.ClusterIP = "10.0.0.10"
+				newSvc.Spec.ClusterIPs = []string{"10.0.0.10"}
+				newSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
+				newSvc.Spec.IPFamilyPolicy = &singleStack
+
+			},
+			numErrs: 3,
+		},
+		{
+			name: "valid: convert to ExternalName",
+			tweakSvc: func(oldSvc, newSvc *core.Service) {
+				singleStack := core.IPFamilyPolicySingleStack
+				oldSvc.Spec.Type = core.ServiceTypeClusterIP
+				oldSvc.Spec.ClusterIP = "10.0.0.10"
+				oldSvc.Spec.ClusterIPs = []string{"10.0.0.10"}
+				oldSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
+				oldSvc.Spec.IPFamilyPolicy = &singleStack
+
+				newSvc.Spec.Type = core.ServiceTypeExternalName
+				newSvc.Spec.ExternalName = "foo"
+			},
+			numErrs: 0,
+		},
+
 		{
 			name: "same ServiceIPFamily",
 			tweakSvc: func(oldSvc, newSvc *core.Service) {
-				ipv4Service := core.IPv4Protocol
 				oldSvc.Spec.Type = core.ServiceTypeClusterIP
-				oldSvc.Spec.IPFamily = &ipv4Service
+				oldSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
 
 				newSvc.Spec.Type = core.ServiceTypeClusterIP
-				newSvc.Spec.IPFamily = &ipv4Service
+				newSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
 			},
 			numErrs: 0,
 		},
+		{
+			name: "same ServiceIPFamily, change IPFamilyPolicy to singleStack",
+			tweakSvc: func(oldSvc, newSvc *core.Service) {
+				oldSvc.Spec.Type = core.ServiceTypeClusterIP
+				oldSvc.Spec.IPFamilyPolicy = nil
+				oldSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
+
+				oldSvc.Spec.IPFamilyPolicy = &singleStack
+				newSvc.Spec.Type = core.ServiceTypeClusterIP
+				newSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
+			},
+			numErrs: 0,
+		},
+		{
+			name: "same ServiceIPFamily, change IPFamilyPolicy singleStack => preferDualStack",
+			tweakSvc: func(oldSvc, newSvc *core.Service) {
+				oldSvc.Spec.Type = core.ServiceTypeClusterIP
+				oldSvc.Spec.IPFamilyPolicy = &singleStack
+				oldSvc.Spec.IPFamilies = []core.IPFamily{core.IPv4Protocol}
+
+				oldSvc.Spec.IPFamilyPolicy = &requireDualStack
```

Comment:
This should be updating newSvc? Also the test name says `singleStack => preferDualStack"` but we're updating the Service to `requireDualStack` here